libs/commons/Textedit.ml

```diff
@@ -33,33 +37,50 @@ type edit_application_result =
       conflicting_edits : t list;
     }
 
+let overlaps (e1 : t) (e2 : t) =
+  (* overlap <=> not (no overlap)
+             <=> not (e1 ends before e2 or e2 ends before e1) *)
+  not (e1.end_ <= e2.start || e2.end_ <= e1.start)
+
+let overlaps_with_an_interval (intervals : t list) (e1 : t) =
+  List.exists (fun e2 -> overlaps e1 e2) intervals
+
+(*
+   Remove overlapping edits by proceeding from left to right in the order
+   the edits are given to us.
+
+   This is an O(N^2) algorithm where N is the number of edits. It should be
+   small enough in practice since N is the number of edits to make in one
+   file. If it turns out this isn't good enough, we'll have to use a more
+   elaborate algorithm. The problem is that we have to apply the fixes
+   in the order in which they're discovered by semgrep rules rather than
+   than in the order in which they appear in the file.
+*)
 let remove_overlapping_edits edits =
-  let rec f edits conflicting_edits = function
-    | e1 :: e2 :: tl ->
-        if e1.end_ > e2.start then
-          let conflicting_edits =
-            (* If the edits are identical, they are not conflicting. We can
-             * apply just one. *)
-            if e1 = e2 then conflicting_edits else e2 :: conflicting_edits
-          in
-          f edits conflicting_edits (e1 :: tl)
-        else
-          let edits = e1 :: edits in
-          f edits conflicting_edits (e2 :: tl)
-    | [ edit ] ->
-        let edits = edit :: edits in
-        (List.rev edits, List.rev conflicting_edits)
-    | [] -> (List.rev edits, List.rev conflicting_edits)
+  let accepted_edits, conflicting_edits =
+    List.fold_left
+      (fun (accepted_edits, conflicting_edits) edit ->
+        if overlaps_with_an_interval accepted_edits edit then
+          (accepted_edits, edit :: conflicting_edits)
+        else (edit :: accepted_edits, conflicting_edits))
+      ([], []) edits
   in
-  f [] [] edits
+  (List.rev accepted_edits, List.rev conflicting_edits)
 
-let apply_edit_to_text text { start; end_; replacement_text; _ } =
+let apply_edit_to_text text ({ start; end_; replacement_text; _ } as edit) =
+  Logs.debug (fun m -> m ~tags "Apply edit %s" (show edit));
   let before = Str.string_before text start in
   let after = Str.string_after text end_ in
   before ^ replacement_text ^ after
 
-let apply_edits_to_text text edits =
-  let edits = List.sort (fun e1 e2 -> e1.start - e2.start) edits in
+let apply_edits_to_text path text edits =
+  (*
+     Don't sort the edits. They must be applied in the order in which they
+     were detected. If two rules report a problem at the same location,
+     the first rule that reports the problem has precedence.
+  *)
```

Comment:
It's fine to deduplicate before applying, but after that we do need to sort them and then apply them in reverse order. Otherwise, edits applied earlier in the file will shift the text later in the file, and subsequent edits will be applied in the wrong place. I suspect this is why the tests are failing.

Reply:
Yes, I found out about it the hard way but the existing comment explaining this was helpful.